tools/toolchains/mbed_toolchain.py

```diff
@@ -129,6 +129,22 @@ class mbedToolchain(with_metaclass(ABCMeta, object)):
 
     PROFILE_FILE_NAME = ".profile"
 
+    # NOTE: GCC_ARM_LTO_WORKAROUND
+    # This is a workaround for the GCC not using the strong symbols from
+    # C files to override the weak symbols from ASM files. This GCC bug is only
+    # present when building with the link-time optimizer (LTO) enabled. For
+    # more details please see:
+    #   https://gcc.gnu.org/bugzilla/show_bug.cgi?id=83967
+    #
+    # This can be fixed by changing the order of object files in the linker
+    # command; objects providing the weak symbols and compiled from assembly
+    # must be listed before the objects providing the strong symbols.
+    # To keep things simple, ALL object files from ASM are listed before
+    # other object files.
+    ASM_OBJ_INFO_LIST_NAME = 'asm_objects'
+    # The name of a JSON file used to store a list of object files compiled from ASM sources.
+    ASM_OBJ_INFO_FILE_NAME = 'asm_obj.json'
+
```

Comment:
This change seems only to affect the gcc compiler. It therefore should only appear in the gcc.py module, otherwise there is a risk that an inadvertent change here would adversely affect other toolchain implementations.

Reply:
I moved all the code to `gcc.py`. @mark-edgeworth, could you review this again, please?